src/Geometry/K3Auto.jl

```diff
@@ -0,0 +1,2059 @@
+
+@doc Markdown.doc"""
+    K3Chamber
+
+The chamber in `S` induced from a Weyl vector in `L`.
+"""
+mutable struct K3Chamber
+  weyl_vector::fmpz_mat
+  # for v in walls, the corresponding half space is defined by the equation
+  # x * gram_matrix(S)*v >= 0, further v is primitive in S (and, in contrast to Shimada, not S^\vee)
+  walls::Vector{fmpz_mat}
+  lengths::Vector{fmpq}  #
+  B::fmpz_mat # QQ-basis consisting of rays #... why do we bother to save this?
+  gramB::fmpz_mat # the basis matrix inferred from the QQ-basis
+  parent_wall::fmpz_mat # for the spanning tree
+  data::BorcherdsCtx
+  fp::Matrix{Int} # fingerprint for the backtrack search
+  #per::Vector{Int}  # permutation
+  fp_diagonal::Vector{Int}
+
+  # TODO: Be more memory efficient and store only the indices for the
+  # basis matrix and the permutation.
+  # I am not sure if storing gram matrix stuff in memory actually increases performance...
+
+  function K3Chamber()
+    return new()
+  end
+end
+
+export chamber, BorcherdsCtx
+
+@doc Markdown.doc"""
+    chamber(data::BorcherdsCtx, weyl_vector::fmpz_mat, [parent_wall::fmpz_mat, walls::Vector{fmpz_mat}])
+
+Return the chamber with the given weyl vector.
+"""
+function chamber(data::BorcherdsCtx, weyl_vector::fmpz_mat, parent_wall::fmpz_mat=zero_matrix(ZZ, 0, 0))
+  D = K3Chamber()
+  D.weyl_vector = weyl_vector
+  D.parent_wall = parent_wall
+  D.data = data
+  return D
+end
+
+function chamber(data::BorcherdsCtx, weyl_vector::fmpz_mat, parent_wall::fmpz_mat, walls::Vector{fmpz_mat})
+  D = K3Chamber()
+  D.weyl_vector = weyl_vector
+  D.parent_wall = parent_wall
+  D.data = data
+  D.walls = walls
+  return D
+end
+
+# needed to create sets of K3Chambers
+function Base.hash(C::K3Chamber)
+  return hash(C.weyl_vector[:,1:rank(C.data.S)])
+end
+
+# Two chambers are equal if and only if their Weyl vectors
+# project to the same point in S
+# By the choice of our coordinates this projection is determined
+# by the first rank(S) coordinates.
+function Base.:(==)(C::K3Chamber, D::K3Chamber)
+  @req C.data===D.data "K3Chambers do not have the same context"
+  return C.weyl_vector[:,1:rank(C.data.S)] == D.weyl_vector[:,1:rank(D.data.S)]
+end
+
+@doc Markdown.doc"""
+    walls(D::K3Chamber) -> Vector{fmpz_mat}
+
+Return the walls of the chamber `D`, i.e. its facets.
+
+The corresponding half space of the wall defined by `v` in `walls(D)` is
+
+```Math
+\{x \in S \otimes \mathbb{R} |  \langle x,v \rangle  \geq 0\}.
+```
+
+`v` is given with respect to the basis of `S` and is primitive in `S`.
+Note that Shimada follows a different convention and takes `v` primitive in `S^\vee`.
+"""
+function walls(D::K3Chamber)
+  if !isdefined(D, :walls)
+    D.walls = _walls_of_chamber(D.data, D.weyl_vector)
+    @assert length(D.walls)>=rank(D.data.S) "$(D.weyl_vector)"
+  end
+  return D.walls
+end
+
+weyl_vector(D::K3Chamber) = D.weyl_vector
+
+@doc Markdown.doc"""
+    rays(D::K3Chamber)
+
+Return the rays of the induced chamber `D`.
```

Comment:
Is `D` the "induced chamber"? Then perhaps just remove the "induced"? Or does the chamber `D` induce some other chamber whose rays are returned? Then this should be explained perhaps?